ext/mysqli/mysqli_nonapi.c

```diff
@@ -355,16 +355,18 @@ void mysqli_common_connect(INTERNAL_FUNCTION_PARAMETERS, bool is_real_connect, b
 
 	mysql->multi_query = 0;
 
-	if (!object) {
-		MYSQLI_RETURN_RESOURCE(mysqli_resource, mysqli_link_class_entry);
-	} else {
+	if (object) {
 		ZEND_ASSERT(instanceof_function(Z_OBJCE_P(object), mysqli_link_class_entry));
 		(Z_MYSQLI_P(object))->ptr = mysqli_resource;
-	}
-	if (!is_real_connect) {
-		return;
-	} else {
 		RETURN_TRUE;
+	} else {
+		MYSQLI_RETURN_RESOURCE(mysqli_resource, mysqli_link_class_entry);
+		if (is_real_connect) {
+			zval_ptr_dtor(return_value);
```

Comment:
I'm wondering if this was missed previously.. We overwrite the return value so `zval_ptr_dtor` should be used, right? Sorry if I confuse things :S

Reply:
The change *looks* right to me, but I don't understand how this has not been an issue before. Shouldn't there be leak warnings?

Reply:
Yeah, I am also puzzled, but I have zero idea why it was not caught :(

Reply:
Wow, this code is seriously confusing. Turns out that `object` is set not just for method calls, but also for real connect calls! So actually `!object && is_real_connect` is impossible, and this is dead code. So you can replace this block with just `return;`.

Reply:
If the code is so confusing that 3 people between themselves can't understand it, then maybe it needs to be refactored.

Reply:
This is what the method should return for all cases (except the special ctor case):
| call                  | object | is_real_connect | in_ctor | return                  |
|-----------------------|--------|-----------------|---------|-------------------------|
| new mysqli            | true   | false           | true    | void                    |
| mysqli_connect        | null   | false           | false   | mysqli\|false           |
| $mysqli->connect      | true   | false           | false   | should be bool, is void |
| $mysqli->real_connect | true   | true            | false   | bool                    |
| mysqli_real_connect   | true   | true            | false   | bool                    |

Reply:
> This is what the method should return for all cases (except the special ctor case):

If we really don't mind if the `return_value` becomes `TRUE` in case of the constructor (like Nikita said), then the implementation suggested by Nikita is fine, I think.